### ISSUE TITLE
Recaptcha V2 just spins

### ISSUE THREAD

**Issue by autohpn** [google/recaptcha#539]
**Issue description**
I have a problem with recaptcha v2. It has stopped working without having changed anything. When checking, it only rotates and does not verify. 

**Environment**

 * OS name and version: APACHE
 * PHP version: 5.3.29
 * Web server name and version: Linux 2.6.32-642.1.1.el6.x86_64
 * `google/recaptcha` version: V2
 * Browser name and version: ALL

**Reproducing the issue**

  * URL (optional): https://www.autohispania.com/contact_us.php

**Comment by stdex** [comment]
This should be conflict with mootools - https://stackoverflow.com/questions/77473778/google-recaptcha-keeps-on-loading   

Compatability bind, that could have worked earlier (https://github.com/google/recaptcha/issues/374#issuecomment-729881795) also stop working.

**Comment by autohpn** [comment]
Thanks for your answer! 

how can i solve this????? I don't found the solution :-(

**Comment by dimov-cz** [comment]
What version of Mootools do you use? 1.6.0 resolves some Array.from related issues.

**Comment by stdex** [comment]
I can confirm problem on mootools 1.4.2

**Comment by autohpn** [comment]
the problem happens with both 1.4.2 and 1.6.

I have managed to solve the problem by not loading mootools in the urls where the captcha loads. It is the only radical way by which I have been able to do it.

**Comment by dimov-cz** [comment]
1.6.0 without comp. layer is working for me

**Comment by dimov-cz** [comment]
The problematic code in compatibility mode is:
/*<1.5compat>*/
Array.from = Array.convert;
/*</1.5compat>*/

and in 1.5.1 and earlier version Array.from implementation which doesn't handle objects like MapIterate and Set.

**Comment by SaintAnd** [comment]
For me it's worked in template index.php after <jdoc:include type="head"/> :
`<?php unset($this->_scripts[$this->baseurl.'/media/system/js/mootools-core.js']); ?>`

**Comment by am-unklown** [comment]
> For me it's worked in template index.php after <jdoc:include type="head"/> : `<?php unset($this->_scripts[$this->baseurl.'/media/system/js/mootools-core.js']); ?>`

it removes core of mootools which can be used somewhere in your site. for example in my case it is used in virtuemart users/orders pages. removing it may lead to other issues.

**Comment by am-unklown** [comment]
> 1.6.0 without comp. layer is working for me

for me it resolves problem with recaptcha but brings problems in admin part of site (e.g. on pages with availability to sort entities: articles, orders, etc). enabling compat layer even with fix form #374 brokes recaptcha.

**Comment by SaintAnd** [comment]
> > For me it's worked in template index.php after <jdoc:include type="head"/> : `<?php unset($this->_scripts[$this->baseurl.'/media/system/js/mootools-core.js']); ?>`
> 
> it removes core of mootools which can be used somewhere in your site. for example in my case it is used in virtuemart users/orders pages. removing it may lead to other issues.

I agree with you. This leads to other problems on the site. Spoilers and sliders break

**Comment by SaintAnd** [comment]
Unfortunately, this is the only option currently used.
<?php if ($_SERVER["REQUEST_URI"] == "/internet-priemnaya/") { 
unset($this->_scripts[$this->baseurl.'/media/system/js/mootools-core.js']);}
 ?>

**Comment by jacksleight** [comment]
Using @sleemanj's [original bind compatibility approach](https://gist.github.com/sleemanj/f076ed2c0b887ab08074b55dad2fd636) and the `Array.from` [hint](https://github.com/google/recaptcha/issues/539#issuecomment-1813155836) from @dimov-cz I've come up with the following which seems to be working for me (mootools 1.4.5):

```js
// Add this BEFORE mootools is loaded

Function.prototype._nativeBind = Function.prototype.bind;
Array._nativeFrom = Array.from;
```

```js
// Add this AFTER mootools is loaded

var isRecaptcha = function() {
	var stack = new Error().stack;
	if (!stack) {
		try {
			throw new Error();
		} catch (e) {
			stack = e.stack;
		}
	}
	return stack.match(/recaptcha/);
}

Function.prototype._mootoolsBind = Function.prototype.bind;
Array._mootoolsFrom = Array.from;

Function.prototype.bind = function (...args) {
	return isRecaptcha() ? this._nativeBind(...args) : this._mootoolsBind(...args);
};
Array.from = function (...args) {
	return isRecaptcha() ? this._nativeFrom(...args) : this._mootoolsFrom(...args);
};
```

I've not tested this thoroughly yet and there could be issues with it! Improvements welcome.

**Comment by pinta83** [comment]
Okay, I have a much elegant solution for this....
Change this part of code
`Array.from = function(item){
	if (item == null) return [];
	return (Type.isEnumerable(item) && typeof item != 'string') ? (typeOf(item) == 'array') ? item : slice.call(item) : [item];
};`
with this
`Array.from = function(item){
	if(new Error().stack.indexOf('recaptcha') >= 0) return [].slice.call(item);
	if (item == null) return [];
	return (Type.isEnumerable(item) && typeof item != 'string') ? (typeOf(item) == 'array') ? item : slice.call(item) : [item];
};`

I can upload patched files if anyone needs....

**Comment by AlexSmetaninGB** [comment]
post the file if it works please

**Comment by pinta83** [comment]
> post the file if it works please

Edited and uploaded in thread above ;)

**Comment by Benderalfa** [comment]
Does not work for Joomla version 3.10
Please look at the files of Joomla version 3.10
Please
[Joomla3.10.zip](https://github.com/google/recaptcha/files/13477969/Joomla3.10.zip)

**Comment by Benderalfa** [comment]
It was in the Joomla cache or in the browser cache. I cleaned both there and there. Works thanks!

**Comment by AlexSmetaninGB** [comment]
thx all ;_)
pinta83-us (dev) dimov-cz (dev) SaintAnd-us (sql) jacksleight-us (j.s) AlexSmetaninGB-ru (dev, bug report)

**Comment by AlexSmetaninGB** [comment]
> thx all ;_) pinta83-us (dev) dimov-cz (dev) SaintAnd-us (sql) jacksleight-us (j.s) AlexSmetaninGB-ru (dev, bug report)

pinta83-WO (dev)
dimov-cz (dev)
SaintAnd- ru (sql) Andrew Gospodin
jacksleight-United Kingdom (j.s) Jack Sleight
AlexSmetaninGB-ru (dev, bug report) AlexSmetanin

**Comment by trinkel** [comment]
Note that you may need to clear the browser cache for this to work.

**Comment by sempostma** [comment]
This works without having to run a script before mootools is loaded:

```javascript
(function fixMooToolsAndRecaptchaMapsIssueIffy() {
// Add this AFTER mootools is loaded

// restore Native Array.from by getting the native implemenation from a new frame.
var frame = document.createElement('iframe');
frame.sandbox = 'allow-same-origin';
document.body.appendChild(frame);
Array._nativeFrom = frame.contentWindow.Array.from
Function.prototype._nativeBind= frame.contentWindow.Function.prototype.bind
frame.remove();
var isRecaptchaOrMaps = function() {
	var stack = new Error().stack;
	if (!stack) {
		try {
			throw new Error();
		} catch (e) {
			stack = e.stack;
		}
	}
	return stack.match(/recaptcha/) || stack.match(/maps\.google\.com/);
}

Function.prototype._mootoolsBind = Function.prototype.bind;
Array._mootoolsFrom = Array.from;

Function.prototype.bind = function (...args) {
	return isRecaptchaOrMaps () ? this._nativeBind(...args) : this._mootoolsBind(...args);
};
Array.from = function (...args) {
	return isRecaptchaOrMaps () ? this._nativeFrom(...args) : this._mootoolsFrom(...args);
};
})();
```
